examples/gogo/app/app.bndrun

```diff
@@ -0,0 +1,12 @@
+index: target/index.xml;name="app"
+
+-standalone: ${index}
+
+-runrequires: osgi.identity;filter:='(osgi.identity=org.osgi.enroute.examples.gogo.command)', \
+ osgi.identity;filter:='(osgi.identity=org.apache.felix.gogo.shell)',\
```

Comment:
It should not be necessary to include requirements for something like the Felix Gogo shell in a bndrun. There should be sufficient information in the example bundle's requirements to pull in everything needed to make it work. For example in the Quickstart tutorial: https://github.com/osgi/osgi.enroute/blob/7d534c287aa6e3ee9391f815d03ca4f81e480356/examples/quickstart/app/app.bndrun#L5

Reply:
So, I should add a requirement to the bundle?
Currently the bundle does not contain a requirement that Gogo is present as it does not depend on any specific part that Gogo provides.

It's also the reason the Gogo runrequires is part of `app` and not only debug.
The app should show the Gogo part and Gogo should not only be part of the debug app.

Reply:
> So, I should add a requirement to the bundle?

Yes

> Currently the bundle does not contain a requirement that Gogo is present as it does not depend on any specific part that Gogo provides.

Typically this is handled through the use of an `osgi.implementation` capability (see the Http Whiteboard and JAX-RS Whiteboard). This ensures that an implementation is provisioned even though there is no direct package or service dependency between the clients. Again, this may involve some feedback to the Gogo team to get them to add the necessary capability.

Reply:
I checked the Gogo bundles. They do not provide any `osgi.implementation` or another (for that approach) useful capability.
As long as Gogo does not provide such stuff, I don't see any change to fulfil that requirements.

Reply:
I am still thinking about the requirement.
I assume mostly a command is added to perform that logic that needs access to internal non exported stuff. At least I hit that situation most of the time. To access non exported packages the commands are "normally" part of specific bundles.
The commands itself are just an addition for someone who wants to use a console to interact with the code.

Shouldn't the bundle requirements also be fulfilled without a command implementation?

Do you really think it would make sense to add a requirement to the bundles manifest because a bundle adds some functions?
Wouldn't it make sense to add this only to the run configuration of that example?

You stated it should be an example that user can use without all the technical details (has my reading be correct?). If everyone adds requirements for a console implementation only because they provided some "additional" command, couldn't it be confuse the consumer of the bundle because a console implementation is necessary to use that bundle in a non interactive runtime?

Reply:
My understanding is that the purpose of this example is to demonstrate how to provide a command in Gogo. In this case the *only* thing that this bundle does is to provide a command and it's pretty useless without it. The result being - this bundle requires Gogo.

A similar situation occurs with the quickstart tutorial. The bundle providing the JAX-RS microservice has a requirement for the JAX-RS Whiteboard, as without this the bundle doesn't really do anything.

> The commands itself are just an addition for someone who wants to use a console to interact with the code.

If this really is the case the commands should be provided in a separate bundle (or a fragment if they need access to internals) so that they can be deployed separately. This separate bundle or fragment should then have a requirement for the Gogo Shell, which it needs.

A big problem in software comes from the poor cohesion of features leading to excessive dependency fanout and a poor understanding of what's actually required to make a module run. We should be helping people to avoid doing that.